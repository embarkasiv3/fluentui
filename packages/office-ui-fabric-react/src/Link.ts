export * from './packages/react-fundamentals/Link';
