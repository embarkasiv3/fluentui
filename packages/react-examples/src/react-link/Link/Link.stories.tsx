import * as React from 'react';
import { Stack, IStackTokens, Text } from '@fluentui/react';
import { Link, LinkProps } from '@fluentui/react-link';

const linksStackTokens: IStackTokens = {
  childrenGap: 10,
};

const LinkExamples = (props: LinkProps) => (
  <Stack tokens={linksStackTokens}>
    <div>
      <Link {...props}>Standalone link</Link>
    </div>
    <div>
      <Link {...props} disabled>
        Standalone disabled link
      </Link>
    </div>
    <div>
      <Link {...props} disabled disabledFocusable>
        Standalone disabled focusable link
      </Link>
    </div>
    <div>
      This is{' '}
      <Link {...props} inline>
        a link
      </Link>{' '}
      used alongside other text content.
    </div>
    <div>
      This is{' '}
      <Link {...props} inline disabled>
        a disabled link
      </Link>{' '}
      used alongside other text content.
    </div>
    <div>
      This is{' '}
      <Link {...props} inline disabled disabledFocusable>
        a disabled focusable link
      </Link>{' '}
      used alongside other text content.
    </div>
  </Stack>
);

const examplesStackTokens: IStackTokens = {
  childrenGap: 20,
};

export const Links = () => (
  <Stack tokens={examplesStackTokens}>
    <Text variant="xLarge">Links with an href render as an anchor</Text>
    <LinkExamples href="https://www.bing.com" />

    <Text variant="xLarge">Links without an href render as a button</Text>
    <LinkExamples />
  </Stack>
);
