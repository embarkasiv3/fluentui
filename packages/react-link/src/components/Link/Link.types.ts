import * as React from 'react';
import { ComponentProps } from '@fluentui/react-utilities';

/**
 * {@docCategory Link}
 */
export type LinkProps = ComponentProps &
  React.AnchorHTMLAttributes<HTMLAnchorElement | HTMLButtonElement | HTMLElement> &
  Omit<React.ButtonHTMLAttributes<HTMLAnchorElement | HTMLButtonElement | HTMLElement>, 'type'> & {
    /**
     * Whether the link is disabled.
     * @defaultvalue false
     */
    disabled?: boolean;

    /**
     * When set, allows the link to be focusable even when it has been disabled. This is used in scenarios where it is
     * important to keep a consistent tab order for screen reader and keyboard users.
     * @defaultvalue false
     */
    disabledFocusable?: boolean;

    /**
     * URL the link points to. If not provided, the link renders as a button (unless that behavior is
     * overridden using `as`).
     */
    href?: string;

    /**
     * If true, changes styling when the link is being used alongside other text content.
     */
    inline?: boolean;

    /**
     * Click handler for the link.
     */
    onClick?: (event: React.MouseEvent<HTMLAnchorElement | HTMLButtonElement | HTMLElement>) => void;

    /**
     * If true, changes the link styling to emphasize that it represents an alternative action.
     * @defaultvalue false
     */
    secondary?: boolean;

    /**
     * Relationship to the linked URL (can be a space-separated list).
     * Most common values are `noreferrer` and/or `noopener`.
     * This prop is only applied if `href` is set.
     */
    rel?: string;

    /**
     * Where to open the linked URL. Common values are `_blank` (a new tab or window),
     * `_self` (the current window/context), `_parent`, and `_top`.
     * This prop is only applied if `href` is set.
     */
    target?: string;

    /**
     * Built-in HTML attribute with different behavior depending on how the link is rendered.
     * If rendered as `<a>`, hints at the MIME type.
     * If rendered as `<button>`, override the type of button (`button` is the default).
     */
    type?: string;

    /** Style tokens */
    // tokens?: RecursivePartial<LinkTokens>;
  };

/**
 * {@docCategory Link}
 */
export interface LinkState extends LinkProps {}

// /**
//  * {@docCategory Link}
//  */
// export type LinkTokens = ColorTokens &
//   FontTokens & {
//     // The text decoration used for the link in its default state.
//     textDecoration?: string;

//     // The set of color tokens that are applied when the link has been visited.
//     visited?: ColorTokenSet;

//     // The text decoration used for the link when it is being focused.
//     focused?: {
//       textDecoration: string;
//     };

//     // The text decoration used for the link when it is being hovered.
//     hovered?: {
//       textDecoration: string;
//     };

//     // The text decoration used for the link when it is being pressed.
//     pressed?: {
//       textDecoration: string;
//     };

//     // The text decoration used for the link when it is in a disabled state.
//     disabled?: {
//       textDecoration: string;
//     };
//   };

// /**
//  * {@docCategory Link}
//  */
// export type LinkVariants<TTokens = LinkTokens> = {
//   root?: TTokens;
// };
