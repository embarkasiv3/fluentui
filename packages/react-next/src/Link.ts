import './version';
export * from '@fluentui/react-link/lib/Link';
