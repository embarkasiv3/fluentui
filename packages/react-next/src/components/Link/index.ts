export * from './Link';
export * from './LinkBase';
export * from './Link.types';
export * from './useLink';
