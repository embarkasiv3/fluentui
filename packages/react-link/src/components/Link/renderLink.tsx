import * as React from 'react';
import { getSlots } from '@fluentui/react-utilities';
import { LinkState } from './Link.types';
import { linkShorthandProps } from './useLink';

/**
 * Defines the render function. Given the state of a Link, renders it.
 */
export const renderLink = (state: LinkState) => {
  const { slots, slotProps } = getSlots(state, linkShorthandProps);

  return <slots.root {...slotProps.root} />;
};
