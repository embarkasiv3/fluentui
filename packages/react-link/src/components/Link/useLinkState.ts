import * as React from 'react';
import { getCode, EnterKey, SpacebarKey } from '@fluentui/keyboard-key';
import { LinkState } from './Link.types';

/**
 * The useLink hook processes the Link draft state.
 * @param state - Link draft state to mutate.
 */
export const useLinkState = (state: LinkState) => {
  const { as, disabled, disabledFocusable, href, onClick, onKeyDown: onKeyDownCallback, type } = state;

  state.onClick = (ev: React.MouseEvent<HTMLAnchorElement | HTMLButtonElement | HTMLElement>) => {
    if (disabled) {
      ev.preventDefault();
    } else {
      onClick?.(ev);
    }
  };

  const onKeyDown = (ev: React.KeyboardEvent<HTMLAnchorElement | HTMLButtonElement | HTMLElement>) => {
    const keyCode = getCode(ev);
    const isEnterOrSpaceKey = keyCode === EnterKey || keyCode === SpacebarKey;
    if (disabled && isEnterOrSpaceKey) {
      ev.preventDefault();
      ev.stopPropagation();
    } else {
      onKeyDownCallback?.(ev);

      if (onClick && isEnterOrSpaceKey) {
        onClick((ev as unknown) as React.MouseEvent<HTMLAnchorElement | HTMLButtonElement | HTMLElement>);
      }
    }
  };

  // Adjust props depending on the root type.
  if (typeof as === 'string') {
    if (as === 'a') {
      state.href = disabled ? undefined : href;
      state.tabIndex = disabled && !disabledFocusable ? undefined : 0;
    } else {
      // Remove the href, rel and target props for all non-anchor elements.
      state.href = undefined;
      state.rel = undefined;
      state.target = undefined;

      // Add the type="button" prop for button elements
      if (as === 'button') {
        state.type = type ? type : 'button';
      }
      // Add keydown event handler for all other elements
      else {
        state.onKeyDown = onKeyDown;
      }
    }
  }
  // Add keydown event handler for all other elements
  else {
    state.onKeyDown = onKeyDown;
  }

  // Add aria attributes
  state['aria-disabled'] = disabled || disabledFocusable;
  state.disabled = as === 'a' || as === 'button' ? disabled && !disabledFocusable : undefined;
};
