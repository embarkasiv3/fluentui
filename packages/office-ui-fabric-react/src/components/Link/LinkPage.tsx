import * as React from 'react';
import { LinkBasicExample } from './examples/Link.Basic.Example';
import { DemoPage } from '../../demo/components/DemoPage';
import { IDemoPageProps } from '../../demo/components/DemoPage.types';
import { LinkStatus } from './Link.checklist';

const LinkBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Link/examples/Link.Basic.Example.tsx') as string;

export const LinkPageProps: IDemoPageProps = {
  title: 'Link',
  componentName: 'Link',
  componentUrl:
    'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Link',
  componentStatus: LinkStatus,
  examples: [
    {
      title: 'Link',
      code: LinkBasicExampleCode,
      view: <LinkBasicExample />
    }
  ],
  propertiesTablesSources: [require<string>('!raw-loader!office-ui-fabric-react/src/components/Link/Link.types.ts')],
  overview: require<string>('!raw-loader!office-ui-fabric-react/src/components/Link/docs/LinkOverview.md'),
  bestPractices: '',
  dos: require<string>('!raw-loader!office-ui-fabric-react/src/components/Link/docs/LinkDos.md'),
  donts: require<string>('!raw-loader!office-ui-fabric-react/src/components/Link/docs/LinkDonts.md'),
  isHeaderVisible: true,
  allowNativeProps: true,
  nativePropsElement: ['a', 'button']
};

export const LinkPage = (props: { isHeaderVisible: boolean }) => <DemoPage {...{ ...LinkPageProps, ...props }} />;
