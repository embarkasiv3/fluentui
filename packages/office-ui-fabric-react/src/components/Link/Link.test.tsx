import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { Link } from './Link';

describe('Link', () => {
  it('renders Link correctly', () => {
    const component = renderer.create(<Link href='#'>I'm a link</Link>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});