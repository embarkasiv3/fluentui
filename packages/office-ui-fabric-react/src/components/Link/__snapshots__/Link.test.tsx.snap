// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link renders Link correctly 1`] = `
<a
  className="ms-Link"
  href="#"
  onClick={[Function]}
  target={undefined}
>
  I'm a link
</a>
`;
